### PR TITLE
[16.0][IMP] shopinvader signin jwt anonymous cart

### PR DESCRIPTION
**Pull request by qgroulard**
Handle anonymous cart at signin:

- If there is an anonymous cart, transfer it to the real partner (merge with the previous partner cart if there is one)
- Remove the anonymous partner
- Delete the anonymous cookie

Discussion threads regarding `/signin` and anonymous carts can be found here: https://github.com/shopinvader/odoo-shopinvader/issues/1428 and https://github.com/shopinvader/odoo-shopinvader/issues/1510

### CODE REVIEW

File: shopinvader_api_signin_jwt/routers/signin.py [hunk 1]
```diff
@@ -28,17 +28,27 @@ def signin(
...
     Return HTTP code 201 if res.partner created (case of the first signin).
+    Transfer anonymous cart and delete anonymous partner if any.
     """
     if not partner:
-        env[
+        partner = env[
             "shopinvader_api_signin_jwt.signin_router.helper"
         ]._create_partner_from_payload(payload)
         response.status_code = status.HTTP_201_CREATED
+    anonymous_partner = env["res.partner"]._get_anonymous_partner__cookie(
```
Comment (simahawk) [inline shopinvader_api_signin_jwt/routers/signin.py:44]: @qgroulard who is supposed to call `_create_anonymous_partner__cookie`?
Shouldn't it happen the 1st time you call a cart endpoint w/o authenticated customer?

Comment (sbidoul) [reply shopinvader_api_signin_jwt/routers/signin.py:44]: @simahawk this is done by the `auth_jwt_authenticated_or_anonymous_partner_autocreate` FastAPI depend. There is an example in #1364 where the cart routes are mounted with that authentication method.

File: shopinvader_sale_cart/models/sale_order.py [hunk 2]
```diff
@@ -83,3 +83,36 @@ def _get_cart_line(self, product_id):
...
+    def _transfer_cart(self, partner_id):
+        """
+        Transfer the current cart to a given partner
+        """
+        self.ensure_one()
+        cart = self._find_open_cart(partner_id)
+        if not cart:
+            cart = self._create_empty_cart(partner_id)
+        cart._update_cart_lines_from_cart(self)
+        self.unlink()
```
Comment (simahawk) [inline shopinvader_sale_cart/models/sale_order.py]: Why deleting the cart?
And what about the UUID? 
If you delete it you should keep they UUID for ext reference.

Comment (qgroulard) [reply shopinvader_sale_cart/models/sale_order.py]: You are right, it doesn't make sense to delete the cart there.

I need to delete the anonymous cart to be able to delete the anonymous partner. I do it in the /signin route now.

File: shopinvader_sale_cart/models/sale_order.py [hunk 3]
```diff
@@ -83,3 +83,35 @@ def _get_cart_line(self, product_id):
...
+                    "product_uom_qty": cart_line.product_uom_qty,
+                }
+                vals.update(self.env["sale.order.line"]._play_onchanges_cart_line(vals))
+                cmd = (0, None, vals)
+            update_cmds.append(cmd)
+        self.write({"order_line": update_cmds})
+
+    def _transfer_cart(self, partner_id):
+        """
+        Transfer the current cart to a given partner
```
Comment (sbidoul) [inline shopinvader_sale_cart/models/sale_order.py]: Maybe expand this docstring. Something like this?
```suggestion
        Copy the lines of this cart to the current cart of a given partner. Merge lines of identical products.
        If the partner does not have a current cart yet, create one.
        This cart is not deleted. It is the responsibility of the caller to remove it if desired.
```

Comment (simahawk) [reply shopinvader_sale_cart/models/sale_order.py]: >  This cart is not deleted. It is the responsibility of the caller to remove it if desired.

this part is very helpful :wink:

Comment (simahawk) [reply shopinvader_sale_cart/models/sale_order.py]: maybe a `delete` option could be useful?

Comment (sbidoul) [reply shopinvader_sale_cart/models/sale_order.py]: Or another method `_transfer_and_unlink_cart`

Comment (sbidoul) [reply shopinvader_sale_cart/models/sale_order.py]: But in the end it's simply a matter of adding an `unlink` line at the call site, so a different method or a delete flag may not ease the life of the caller that much.

Comment (qgroulard) [reply shopinvader_sale_cart/models/sale_order.py]: Thanks for the input.

I like to keep it simple, I think this is not necessary to add a delete option or a `_transfer_and_unlink_cart()` method. As @sbidoul said, this is just a matter of adding an `unlink` after the call.

I have added the suggested docstring.

Comment (simahawk) [reply shopinvader_sale_cart/models/sale_order.py]: I'm fine w/ the proposed solution, however... I don't really see the point for keeping this cart.
Especially in the ctx of the anon partner you MUST delete the cart.
As this meth is called "transfer cart" where do you see a use for keeping the old cart?

Comment (qgroulard) [reply shopinvader_sale_cart/models/sale_order.py]: I don't know, you tell me :joy: I thought you had a use case for that, as you started this discussion with this comment:

> Why deleting the cart?
> And what about the UUID?
> If you delete it you should keep they UUID for ext reference.

For me personally, the only use I have of this method is in the context of anonymous partners.  But I like the idea to keep this method atomic and let people do what they want with the cart.

**Review by lmignon — CHANGES_REQUESTED** [review]
@qgroulard Can you add some doc on how to enable this functionality when defining your fastapi app (the required addon, how you mount the router and define the authentication policy, ...)

**Comment by lmignon** [comment]
Documentation will be added into another PR

**Review by lmignon — APPROVED** [review]

**Pull request merged**